2016/exercises/final_project/Gabriel-Adrian/main.cpp

```diff
@@ -0,0 +1,519 @@
+//   4-polytopes with 8 vertices”.
+//
+// Dependencies:
+//   GNU Multiple Precision Arithmetic Library
+//   Parma Polyhedra Library
+//
+#include <cstdint>
+#include <array>
+#include <vector>
+
+#include <gmpxx.h>
+#include <ppl.hh>
+namespace ppl = Parma_Polyhedra_Library;
+
+constexpr auto MAX_ITER = 1000000;
+
+// Basis for R²
+const auto X = ppl::Variable(0);
+const auto Y = ppl::Variable(1);
+
+// A point in R²
+struct Point
+{
+    mpq_class x, y;
+
+    Point(mpq_class x, mpq_class y) : x(x), y(y) {}
+
+    // Return a generator to be used with PPL (convenience)
+    ppl::Generator generator() const
+    {
+        auto nx = x.get_num();
+        auto dx = x.get_den();
+        auto ny = y.get_num();
+        auto dy = y.get_den();
+
+        mpz_class z = dx*dy;
+        return ppl::Generator::point(nx*dy*X + ny*dx*Y, z);
+    }
+};
+
+// A point in a Gale diagram
+struct GalePt
+{
+    enum Sign { Positive, Negative };
+
+    Point pt;
+    Sign sign;
+};
+
+// Representation of a Gale diagram
+struct GaleDiagram
+{
+    std::vector<GalePt> points;
+};
+
+// Stores the facets of a polytope.
+struct CombinatorialType
+{
+    // Compact representation of a subset of a given set of size tsize
+    // tsize is never larger than 8 in our case
+    struct PointSet
+    {
+        using Bits = std::uint32_t;
+        Bits bits;
+        size_t tsize; // total size
+
+        // Check for inclusion
+        bool contained_in(CombinatorialType::PointSet other) const
+        {
+            return (bits & other.bits) == bits;
+        }
+
+        // Test for membership of an element
+        bool contains(size_t i) const
+        {
+            return bits & (1 << i);
+        }
+
+        // Lexicographical ordering for sorting, etc
+        struct TotalOrder
+        {
+            bool operator()(PointSet a, PointSet b) const { return a.bits < b.bits; }
+        };
+    };
+
+    using VertId = size_t;
+    using Facet = PointSet;
+
+    size_t n_vert;
+    std::vector<Facet> facets; // sorted by TotalOrder
+};
+
+bool operator==(CombinatorialType::PointSet a, CombinatorialType::PointSet b)
+{
+    return a.tsize == b.tsize && a.bits == b.bits;
+}
+bool operator!=(CombinatorialType::PointSet a, CombinatorialType::PointSet b)
+{
+    return !(a == b);
+}
+
+
+// Obtain the line through two points
+ppl::Linear_Expression mk_line(Point p, Point q)
+{
+    mpq_class x3 = p.y - q.y;
+    mpq_class y3 = q.x - p.x;
+    mpq_class z3 = p.x*q.y - p.y*q.x;
+
+    mpz_class d1 = x3.get_den();
+    mpz_class d2 = y3.get_den();
+    mpz_class d3 = z3.get_den();
+
+    return d2*d3*x3.get_num()*X + d1*d3*y3.get_num()*Y + d1*d2*z3.get_num();
+}
+
+// Return whether the diagram is a affine Gale diagram of a polytope.
+template <bool Neighborly>
+bool is_gale_diagram_of_polytope(const GaleDiagram& diagram)
+{
+    const auto& points = diagram.points;
+
+    for (size_t i = 0; i < points.size(); i++) {
+        auto p = points[i].pt;
```

Comment:
Thanks for this one! I have no idea about the internals of gmp classes, is this creating a full copy of the contents of `mpq_class` or do they use any sharing?